### ISSUE TITLE
Drag bug in grouptask highchart gantt

### ISSUE THREAD

**Issue by MichealWind** [highcharts/highcharts#11486]
Expected behaviour
In a grouptask highchart gantt, when hide some parent item, and drag some other point, It will be fine.

Actual behaviour
Many points disappeared. And I can not interact with gantt.

Uncaught TypeError: Cannot read property 'push' of undefined
at highcharts-gantt.src.js:43547
at Array.forEach ()
at highcharts-gantt.src.js:43546
at Array.forEach ()
at l (highcharts-gantt.src.js:43542)
at Object.getTree (highcharts-gantt.src.js:43636)
at B (highcharts-gantt.src.js:44888)
at highcharts-gantt.src.js:44962
at Array.forEach ()

Live demo with steps to reproduce
https://jsfiddle.net/fav2Lwkb/

hide "Product launch" parent item
drag some item in "test drag"

Product version
Highchart-gantt 7.1.2

Affected browser(s)
all

**Comment by pawelfus** [comment]
Thanks @MichealWind for setting up the ticket with demo!

**Comment by FilipeCosta06** [comment]
This is once again related to how "hidden data" are actually removed from some arrays that are looped on during an `update`. In better words:
> Looks like when a breaker is added (with the toggleCollapse event), points are being deleted from the Series, not just hidden (https://github.com/highcharts/highcharts/blob/master/js/modules/broken-axis.src.js#L345).
> 
> But then, when updateData gets called, the comparison is between points and new data (https://github.com/highcharts/highcharts/blob/master/js/parts/Series.js#L3227).
> 
> Would the solution be to compare data with data, and not data with points? I don't know the codebase enough to be sure about this.

_Originally posted by @anarelion in https://github.com/highcharts/highcharts/issues/9980#issuecomment-457264514_



I had few encounters with this problem (see https://github.com/highcharts/highcharts/issues/11231#issuecomment-512740022 for some context):

> **Internal Note:** This is caused by `series.options.data` is not updated due to [`updateData` can not update points that does not exist in the `series.points` array](https://github.com/highcharts/highcharts/blob/91df181bb32a3d8235d9e83ea527893d55d9df17/ts/parts/Series.ts#L3345-L3375) and that [`options.data` is later set to `series.options.data`](https://github.com/highcharts/highcharts/blob/91df181bb32a3d8235d9e83ea527893d55d9df17/ts/parts/Dynamics.ts#L1417)


_Originally posted by @jon-a-nygaard in https://github.com/highcharts/highcharts/issues/11231#issuecomment-512740022_

In this case, since I was updating manually I had the opportunity to set `series.keys = true` to

> cause Series.update to skip the logic that tries to reuse the existing points, which includes logic that causes the issue, and just create new points.

But I don't think this can be applied here, unless we somehow override the `drop event` and apply this band aid fix: here's my [attempt](https://jsfiddle.net/y6gnd0bj/). **Let's get weird.** _(Please open your console)_

**IT WORKS !** _(at least for me)_ but.... **only once**.
- In the console, initially, we have 8 points.
- After one drop, we still have 8.
- After another one we then have 9 (not even the Point that got dragged ! I moved "Create unit tests" and "Layout" and "Graphics" get duplicated while the last Point gets removed...) **and the visual bugs start occurring**, but without throwing an error. In my case, the last Point disappears, then a lot of shenanigans occur during subsequent drops.

I referenced a similar duplication bug recently https://github.com/highcharts/highcharts/issues/11439 but have yet to receive an answer (and this bug seems hella weird so I'm not getting my hopes up ahah). 

Also, @pawelfus, I have 2 questions for you:

- Is it possible to achieve sub-tasking by ourselves, without relying on `uniqueNames`, by setting `yAxis.categories` etc... ? I need both sub-tasks and multiple tasks per line, while also being able to manipulate and update things dynamically (kinda like a more classic Scheduler library). As such, I'd like to explore any and every options before giving up ahah
- When using `uniqueNames`, dragging only works horizontally even though `draggableY` is set to `true`. Is this intended (thus my previous question) or should I **open another issue** for that so as to not spam this one ?

**Comment by pawelfus** [comment]
Hi @FilipeCosta06 

Thanks for asking, but I think @jon-a-nygaard can provide you much more detailed answer in your case. Anyway, let me try! 

- remove `uniqueNames` - yes, you can remove those, but (@jon-a-nygaard must confirm here) collapsible subtasks are not gonna be interactive, take a look: https://jsfiddle.net/BlackLabel/4k916nyp/ - simply added categories + y-value to a task. Just like in our official demo: https://www.highcharts.com/gantt/demo/interactive-gantt
- drag&drop and `uniqueNames` - the config seems to be wrong, I guess it's copied from another demo, just cleaned up a little and: https://jsfiddle.net/BlackLabel/4k916nyp/

**Comment by FilipeCosta06** [comment]
Excuse me @pawelfus, I wasn't precise enough in my questions _(and I will continue here instead of creating an issue, unless told otherwise)_.

When I say `uniqueNames` I mean "having subtasks" (groups in the yAxis that you can `collapse`/`expand` by clicking on the `parent` task's `tick`).

Also, yes, the demo **are from another demo**. I was trying to make `dragY` work when using `uniqueNames` (i.e., sub-tasks/"collapsable" groups). The reason your last demo (in the edit) doesn't work is because there is my `drop event` that only updates X (`start` and `end`) values, but if you comment it [there is no problem](https://jsfiddle.net/d1j3v6mL/).

Also, do you, or @jon-a-nygaard, know of a function made to "find a yAxis category by Y coordinates" ? Because
- The `drop event` contains the `event.chartY` position of the mouse drop.
- Ticks in `chart.yAxis[0].ticks` have a `label` attribute that, itself, has a `xy` one. 

As such, on `drop event` I could loop on `yAxis.ticks` and look for the smallest absolute value of `event.chartY - yAxis.ticks[i].xy.y` (and handle the case where the drop happens perfectly between two ticks...) in order to determine which is the closest, i.e. the tick where the Point was tried to be dropped and use this information to update the category of the Point (as a reminder, in [my demo](https://jsfiddle.net/y6gnd0bj/), I have sub-tasks and, as such, the `dragY` doesn't work). But I feel like there must be a better way of going about it.

**Comment by pawelfus** [comment]
Ah, okay, thanks for the explanation! I' not sure if I'm following you, but in general, to get x/y position on the chart, use `chart.yAxis[0].toValue(event.chartY)`. You will get exact value, which should be rounded - this way you will get category/tick index.

**Comment by FilipeCosta06** [comment]
**tl;\[won't\]r:** 
1) When all ticks are expanded:
- X drag works fine.
- Y drag fixed-ish
    - don't move a "parent" down into one of its children (because this makes no sense. We'd have to loop and ascend all nodes, whose parent is the one we are moving, 1 level and updating their `parent`) 
    - don't move anything to an empty line as it will not have a point whose `parent` I extract. _Fixable by looping on yAxis's nodes backwards starting from the targeted tick and looking for a parent node i.e. a smaller `depth` (and, obviously since we're going backwards, a smaller `pos`)._
2) When at least 1 Tick is collapsed:
- X Drag same as before (only works once, then Points get removed/duplicated)
- Y Drag same works with the fix, but only once


Hello again, I am officially dense, I hadn't thought of rounding `yAxis.toValue()`'s result and I assumed it was just "luck" that the integer value was often (_always_) the same as the "drop event's target yAxis category/tick".

Anyway, [_**it\*** works_](https://jsfiddle.net/9vn4L01p/2/) (-ish).

### \* The issue(s)
When using "grouped/sub" tasks in a Gantt (i.e. [`yAxis.uniqueNames=true`](https://api.highcharts.com/gantt/yAxis.uniqueNames) and [`data.parent`](https://api.highcharts.com/gantt/series.gantt.data.parent)):
1) If a single tick is `collapsed` (in the demo above, the `Design` tick is collapsed, hidding "Layout" and "Graphics", it causes errors.
2) It is **not possible to drag vertically** (change `yCategory`, or rather `name` because of `uniqueNames`), no matter the collapsed state of the yAxis ticks.

### The solution(s)
1) Band-aid fix, and **only for X drags**: when updating `series`, include the `series.keys` parameter as such:
```
chart.update({
    series: [{
        keys: chart.series[0].options.keys || true,    // Include this
        data: ...
        ...
    }]
});
```
In order to:

> > cause Series.update to skip the logic that tries to reuse the existing points, which includes logic that causes the issue, and just create new points.
> 
> Originally posted by @jon-a-nygaard in #11231 (comment)

Based on a previous Issue and my recent exchanges with @jon-a-nygaard (mentioned in my initial comment above), the `update` function seems to be only on the visible `data`/`ticks`/etc... while using `userOption`'s value, some of which are currently not visible.

> > Internal Note: This is caused by series.options.data is not updated due to updateData can not update points that does not exist in the series.points array and that options.data is later set to series.options.data
> 
> Originally posted by @jon-a-nygaard in #11231 (comment)


2) Using `plotOptions.series.point.events.drop`, we implement our own custom update. The function's 
`event` and `this` (referencing the data, i.e point/task, that was just drag-dropped) contain all we need.
- For X, just use `event.newPoint.start` and `event.newPoint.end`
- For Y, all we have are target coordinates, meaning we need to
    1) Calculate the targeted category/tick index
        Using `Math.round(this.series.yAxis.toValue(event.chartY)) // newY =`
    2) Use this index to find the associated `name`
        `this.series.yAxis.categories[newY]`
    3) Change `parent`'s value (this allows drag from a group to another)
        `this.series.yAxis.mapOfPosToGridNode[newY].nodes[0].parent`
    4) _When calling `chart.update()`, I had to include `yAxis: { uniqueNames: true }` to force the update of the ticks, even when already using the "`series.keys` band-air", AND because I have no purpose for empty lines **AND** this would make my fix bug too as I did not include any check when calling the previous line (_iii._)_ 

Also, don't forget to `return false`, else the "regular" drop event handler will be called.

_I think that's all..._
